[OMPIRBuilder][debug] Fix debug info for variables in target region.

abidh opened the pull request:
When a new function is created to handle OpenMP target region, the variables used in it are passed as arguments. The scope and the location of these variable contains still point to te parent function of the target region. Such variables will fail in Verifier as the scope of the variables will be different from the containing functions.

Currently, flang is the only user of createOutlinedFunction and it does not generate any debug data for the the variables in the target region to avoid this error. When this PR is in, we should be able to remove this limit in the flang (and anyother client) and have the better debug experience for the target region.

This PR changes the location and scope of the variables in the target region to point to correct entities. It is similar to what fixupDebugInfoPostExtraction does for CodeExtractor. I initially tried to re-use that function but found quickly that it would require quite a bit of re-factoring and additions before it could be used. It was much simpler to make the changes locally.

abidh committed:
[OMPIRBuilder][debug] Fix debug info for variables in target region.

When a new function is created to handle OpenMP target region, the
variables used in it are passed as arguments. The scope and the location
of these variable contains still point to te parent function of the
target region. Such variables will fail in Verifier as the scope of the
variables will be different from the containing functions.

Currently, flang is the only user of createOutlinedFunction and it does
not generate any debug data for the the variables in the target region
to avoid this error. When this PR is in, we should be able to remove
this limit in the flang (and anyother client) and have the better
debug experience for the target region.

This PR changes the location and scope of the variables in the target
region to point to correct entities. It is similar to what
fixupDebugInfoPostExtraction does for CodeExtractor. I initially tried
to re-use that function but found quickly that it would require quite
a bit of re-factoring and additions before it could be used. It was
much simpler to make the changes locally.

Dinistro commented on mlir/test/Target/LLVMIR/omptarget-debug-var-1.mlir:
I would argue that this should be investigated. Working with splits to group similar tests into the same files has shown to be very valuable.

abidh commented:
Polite ping.

abidh commented:
Ping. I would appreciate if someone can spare some time to review this.

Meinersbur commented:
Testing for LLVM components should be done in `llvm/tests` not `mlir`.

abidh commented:
> Testing for LLVM components should be done in `llvm/tests` not `mlir`.

Hi @Meinersbur 

Thanks for taking a look. I see that we routinely add test in `mlir/test/Target/LLVMIR` for changes in `OMPIRBuilder.cpp` that relate to MLIR->LLVMIR translation as this PR is. I am not sure how that test can be moved into `llvm/tests`.

I will investigate if this scenario could be created in a `llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp` so that we can also add a test there but it seems difficult to me on an initial look.

Please let me know if the changes otherwise look reasonable.

jsjodin commented on llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:
```diff
@@ -6902,6 +6907,67 @@ static Expected<Function *> createOutlinedFunction(
   for (auto Deferred : DeferredReplacement)
     ReplaceValue(std::get<0>(Deferred), std::get<1>(Deferred), Func);
 
+  DenseMap<const MDNode *, MDNode *> Cache;
```
All this code looks pretty self-contained. Can this be placed in a helper function instead?

abidh replied:
Thanks for reviewing it. I have moved the code to a separate helper function.

abidh committed:
Handle review comments.

Move the variable update to a separate helper function.

abidh committed:
Handle review comments.

There was code duplicated to handle DbgVariableRecord and
DbgVariableIntrinsic. That duplication has been removed. Also checked
if there is valid DISubprogram right at the start of the function.

abidh committed:
Use simpler way to get Module handle.

abidh committed:
Improve handling of privatized variables.

abidh committed:
Fix clang-format issue.

jsjodin reviewed (approved):
LGTM
